File: swift_reference/Views/HabitCheckInCardView.swift
import SwiftUI

struct HabitCheckInCardView: View {
    let habit: Habit
    let viewModel: HabitViewModel
    let targetDate: Date
    let isRetroactive: Bool

    @State private var showPulse: Bool = false
    @State private var isCompleted: Bool = false
    @State private var currentValue: Double = 0
    @State private var showDetail: Bool = false

    @State private var celebrationMilestone: Milestone?
    @State private var milestoneService = MilestoneService()
    @State private var completionVerse: Scripture?

    @Environment(\.storeViewModel) private var store

    private var isPremium: Bool {
        store?.isPremium ?? false
    }

    private let calendar = Calendar.current

    private var dayName: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "EEEE"
        return formatter.string(from: targetDate)
    }

    var body: some View {
        VStack(spacing: 16) {
            Button {
                showDetail = true
            } label: {
                HStack(spacing: 12) {
                    ZStack {
                        Circle()
                            .fill(
                                RadialGradient(
                                    colors: [
                                        habitAccentColor.opacity(isCompleted ? 0.35 : 0.12),
                                        habitAccentColor.opacity(isCompleted ? 0.15 : 0.04)
                                    ],
                                    center: .center,
                                    startRadius: 0,
                                    endRadius: 22
                                )
                            )
                            .frame(width: 44, height: 44)

                        Image(systemName: habit.habitCategory.iconName)
                            .font(.system(size: 18))
                            .foregroundStyle(isCompleted ? habitAccentColor : habitAccentColor.opacity(0.7))
                            .symbolEffect(.bounce, value: showPulse)
                    }

                    VStack(alignment: .leading, spacing: 3) {
                        Text(habit.name)
                            .font(.system(.headline, design: .serif))
                            .foregroundStyle(.primary)

                        Text(habit.purposeStatement)
                            .font(.caption)
                            .foregroundStyle(.secondary)
                            .lineLimit(1)
                    }

                    Spacer()

                    if isCompleted {
                        Image(systemName: "checkmark.circle.fill")
                            .font(.title2)
                            .foregroundStyle(TributeColor.golden)
                    } else {
                        Image(systemName: "chevron.right")
                            .font(.caption)
                            .foregroundStyle(.secondary.opacity(0.4))
                    }
                }
            }
            .buttonStyle(.plain)

            trackingContent

            if let verse = completionVerse, isCompleted {
                VStack(spacing: 4) {
                    Text("\u{201C}\(verse.text)\u{201D}")
                        .font(.system(.caption, design: .serif))
                        .italic()
                        .foregroundStyle(TributeColor.softGold.opacity(0.6))
                        .multilineTextAlignment(.center)
                        .lineSpacing(2)
                    Text(verse.reference)
                        .font(.caption2)
                        .foregroundStyle(TributeColor.golden.opacity(0.4))
                }
                .padding(.top, 4)
                .transition(.opacity)
            }
        }
        .tributeCard()
        .overlay {
            if showPulse {
                GoldenPulseView(dimmed: isRetroactive)
                    .allowsHitTesting(false)
            }
        }
        .sheet(isPresented: $showDetail) {
            NavigationStack {
                HabitDetailView(habit: habit)
                    .toolbar {
                        ToolbarItem(placement: .topBarTrailing) {
                            Button("Done") { showDetail = false }
                                .foregroundStyle(TributeColor.golden)
                        }
                    }
            }
            .preferredColorScheme(.dark)
        }
        .fullScreenCover(item: $celebrationMilestone) { milestone in
            MilestoneCelebrationView(milestone: milestone) {
                celebrationMilestone = nil
            }
            .preferredColorScheme(.dark)
        }
        .onAppear { refreshState() }
        .onChange(of: targetDate) { _, _ in refreshState() }
        .sensoryFeedback(.success, trigger: showPulse)
    }

    private func refreshState() {
        isCompleted = habit.isCompleted(on: targetDate)
        currentValue = habit.entry(for: targetDate)?.value ?? 0
        if isCompleted {
            completionVerse = ScriptureLibrary.completionVerse(for: habit.habitCategory, on: targetDate, isPremium: isPremium)
        } else {
            completionVerse = nil
        }
    }

    @ViewBuilder
    private var trackingContent: some View {
        switch habit.habitTrackingType {
        case .checkIn:
            checkInContent
        case .abstain:
            abstainContent
        case .timed:
            timedContent
        case .count:
            countContent
        }
    }

    private var checkInContent: some View {
        Group {
            if !isCompleted {
                Button {
                    completeCheckIn()
                } label: {
                    HStack(spacing: 6) {
                        Image(systemName: "plus.circle.fill")
                            .font(.subheadline)
                        Text("Check in")
                    }
                    .tributeButton()
                }
            } else {
                HStack {
                    Text(isRetroactive ? "Given on \(dayName)" : "Given to God today")
                        .font(.system(.subheadline, design: .serif))
                        .foregroundStyle(TributeColor.sage)
                    Spacer()
                    Text("\(habit.totalCompletedDays()) total days")
                        .font(.caption)
                        .foregroundStyle(TributeColor.softGold.opacity(0.6))
                }
            }
        }
    }

    private var abstainContent: some View {
        Group {
            if !isCompleted {
                Button {
                    completeCheckIn()
                } label: {
                    HStack(spacing: 6) {
                        Image(systemName: "shield.fill")
                            .font(.subheadline)
                        Text(isRetroactive ? "Were you strong on \(dayName)?" : "Stayed strong today")
                    }
                    .tributeButton(color: TributeColor.sage)
                }
            } else {
                HStack {
                    Label(isRetroactive ? "Strong on \(dayName)" : "Strong today", systemImage: "shield.fill")
                        .font(.system(.subheadline, design: .serif))
                        .foregroundStyle(TributeColor.sage)
                    Spacer()
                    Text("\(habit.totalCompletedDays()) clean days")
                        .font(.caption)
                        .foregroundStyle(TributeColor.softGold.opacity(0.6))
                }
            }
        }
    }

    private var timedContent: some View {
        VStack(spacing: 12) {
            ZStack {
                Circle()
                    .stroke(Color.white.opacity(0.06), lineWidth: 6)

                Circle()
                    .trim(from: 0, to: habit.dailyTarget > 0 ? min(currentValue / habit.dailyTarget, 1.0) : 0)
                    .stroke(
                        isCompleted ? TributeColor.golden : TributeColor.softGold,
                        style: StrokeStyle(lineWidth: 6, lineCap: .round)
                    )
                    .rotationEffect(.degrees(-90))
                    .animation(.easeInOut(duration: 0.5), value: currentValue)

                VStack(spacing: 2) {
                    Text("\(Int(currentValue))")
                        .font(.system(.title2, design: .rounded, weight: .bold))
                        .foregroundStyle(TributeColor.golden)
                    Text("/ \(Int(habit.dailyTarget)) min")
                        .font(.caption2)
                        .foregroundStyle(.secondary)
                }
            }
            .frame(width: 90, height: 90)

            HStack(spacing: 10) {
                ForEach([5, 10, 15], id: \.self) { minutes in
                    Button {
                        addMinutes(Double(minutes))
                    } label: {
                        Text("+\(minutes)")
                            .font(.caption.weight(.semibold))
                            .foregroundStyle(TributeColor.softGold)
                            .padding(.horizontal, 16)
                            .padding(.vertical, 8)
                            .background(TributeColor.surfaceOverlay)
                            .clipShape(.rect(cornerRadius: 8))
                            .overlay(
                                RoundedRectangle(cornerRadius: 8)
                                    .strokeBorder(TributeColor.cardBorder, lineWidth: 0.5)
                            )
                    }
                }
            }
        }
    }

    private var countContent: some View {
        VStack(spacing: 12) {
            HStack(spacing: 4) {
                Text("\(Int(currentValue))")
                    .font(.system(.title3, design: .rounded, weight: .bold))
                    .foregroundStyle(TributeColor.golden)
                Text("/ \(Int(habit.dailyTarget))")
                    .font(.system(.title3, design: .rounded, weight: .bold))
                    .foregroundStyle(.secondary)
                Text(habit.targetUnit)
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
            }

            HStack(spacing: 20) {
                Button {
                    updateCount(max(0, currentValue - 1))
                } label: {
                    Image(systemName: "minus.circle.fill")
                        .font(.title2)
                        .foregroundStyle(.secondary)
                }

                Button {
                    updateCount(currentValue + 1)
                } label: {
                    Image(systemName: "plus.circle.fill")
                        .font(.title2)
                        .foregroundStyle(TributeColor.golden)
                }
            }

            if isCompleted {
                Text("Target reached \u{2014} \(Int(habit.totalValue())) total")
                    .font(.caption)
                    .foregroundStyle(TributeColor.sage)
            }
        }
    }

    private var habitAccentColor: Color {
        switch habit.habitTrackingType {
        case .abstain: return TributeColor.sage
        default: return TributeColor.golden
        }
    }

    private func completeCheckIn() {
        let previousTotal = Double(habit.totalCompletedDays())
        withAnimation(.easeInOut(duration: 0.5)) {
            showPulse = true
            isCompleted = true
        }
        viewModel.checkInHabit(habit, on: targetDate, retroactive: isRetroactive)
        withAnimation(.easeInOut(duration: 0.5).delay(0.8)) {
            completionVerse = ScriptureLibrary.completionVerse(for: habit.habitCategory, on: targetDate, isPremium: isPremium)
        }

        if !isRetroactive {
            let newTotal = previousTotal + 1
            if let milestone = milestoneService.checkForNewMilestone(habit: habit, previousValue: previousTotal, newValue: newTotal) {
                Task {
                    try? await Task.sleep(for: .seconds(1.8))
                    withAnimation { showPulse = false }
                    try? await Task.sleep(for: .seconds(0.3))
                    celebrationMilestone = milestone
                }
                return
            }
        }

        Task {
            try? await Task.sleep(for: .seconds(1.5))
            withAnimation { showPulse = false }
        }
    }

    private func addMinutes(_ minutes: Double) {
        let previousTotal = habit.totalValue()
        let newValue = currentValue + minutes
        withAnimation(.easeInOut(duration: 0.3)) {
            currentValue = newValue
            if newValue >= habit.dailyTarget && !isCompleted {
                isCompleted = true
                showPulse = true
            }
        }
        viewModel.updateTimedEntry(habit, minutes: newValue, on: targetDate, retroactive: isRetroactive)

        if !isRetroactive {
            let newTotal = previousTotal + minutes
            if let milestone = milestoneService.checkForNewMilestone(habit: habit, previousValue: previousTotal, newValue: newTotal) {
                Task {
                    try? await Task.sleep(for: .seconds(1.8))
                    withAnimation { showPulse = false }
                    try? await Task.sleep(for: .seconds(0.3))
                    celebrationMilestone = milestone
                }
                return
            }
        }

        if showPulse {
            Task {
                try? await Task.sleep(for: .seconds(1.5))
                withAnimation { showPulse = false }
            }
        }
    }

    private func updateCount(_ count: Double) {
        let previousTotal = habit.totalValue()
        let diff = count - currentValue
        withAnimation(.easeInOut(duration: 0.3)) {
            currentValue = count
            if count >= habit.dailyTarget && !isCompleted {
                isCompleted = true
                showPulse = true
            }
        }
        viewModel.updateCountEntry(habit, count: count, on: targetDate, retroactive: isRetroactive)

        if !isRetroactive && diff > 0 {
            let newTotal = previousTotal + diff
            if let milestone = milestoneService.checkForNewMilestone(habit: habit, previousValue: previousTotal, newValue: newTotal) {
                Task {
                    try? await Task.sleep(for: .seconds(1.8))
                    withAnimation { showPulse = false }
                    try? await Task.sleep(for: .seconds(0.3))
                    celebrationMilestone = milestone
                }
                return
            }
        }

        if showPulse {
            Task {
                try? await Task.sleep(for: .seconds(1.5))
                withAnimation { showPulse = false }
            }
        }
    }
}

File: lib/views/habits/habit_check_in_card_view.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../../models/habit.dart';
import '../../models/scripture.dart';
import '../../providers/habit_provider.dart';
import '../../providers/store_provider.dart';
import '../../services/milestone_service.dart';
import '../../theme/app_theme.dart';
import '../circles/sos_view.dart';
import '../shared/golden_pulse_view.dart';
import '../shared/milestone_celebration_view.dart';
import '../shared/tribute_paywall_view.dart';
import 'habit_detail_view.dart';

class HabitCheckInCardView extends StatefulWidget {
  final Habit habit;
  final DateTime targetDate;
  final bool isRetroactive;

  const HabitCheckInCardView({
    super.key,
    required this.habit,
    required this.targetDate,
    this.isRetroactive = false,
  });

  @override
  State<HabitCheckInCardView> createState() => _HabitCheckInCardViewState();
}

class _HabitCheckInCardViewState extends State<HabitCheckInCardView> {
  static const _milestoneService = MilestoneService.instance;

  bool _showPulse = false;
  bool _isCompleted = false;
  double _timedMinutes = 0;
  double _countValue = 0;
  Scripture? _completionVerse;
  Milestone? _celebrationMilestone;

  Habit get _habit => widget.habit;
  DateTime get _targetDate => widget.targetDate;

  @override
  void initState() {
    super.initState();
    _refreshState();
  }

  @override
  void didUpdateWidget(HabitCheckInCardView old) {
    super.didUpdateWidget(old);
    if (old.targetDate != widget.targetDate) {
      _refreshState();
    }
  }

  void _refreshState() {
    final entry = _habit.entryFor(_targetDate);
    _isCompleted = entry?.isCompleted ?? false;
    _timedMinutes = entry?.value ?? 0;
    _countValue = entry?.value ?? 0;
    if (_isCompleted) {
      // isPremium read deferred to build time to avoid context-in-initState issues
      _completionVerse = ScriptureLibrary.completionVerse(_habit.habitCategory, _targetDate, isPremium: false);
    }
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    if (_isCompleted) {
      final isPremium = context.read<StoreProvider>().isPremium;
      _completionVerse = ScriptureLibrary.completionVerse(_habit.habitCategory, _targetDate, isPremium: isPremium);
    }
  }

  Future<void> _checkIn() async {
    final provider = context.read<HabitProvider>();
    final storeProvider = context.read<StoreProvider>();
    final previousTotal = _habit.totalCompletedDays().toDouble();
    setState(() {
      _showPulse = true;
      _isCompleted = true;
    });
    await provider.checkInHabit(_habit, date: _targetDate, retroactive: widget.isRetroactive);
    if (!mounted) return;
    final isPremium = storeProvider.isPremium;
    setState(() {
      _completionVerse = ScriptureLibrary.completionVerse(_habit.habitCategory, _targetDate, isPremium: isPremium);
    });
    if (!widget.isRetroactive) {
      final newTotal = previousTotal + 1;
      final milestone = _milestoneService.checkForNewMilestone(
        _habit,
        previousValue: previousTotal,
        newValue: newTotal,
      );
      if (milestone != null) {
        await Future.delayed(const Duration(milliseconds: 1800));
        if (mounted) setState(() { _showPulse = false; _celebrationMilestone = milestone; });
        return;
      }
    }
    await Future.delayed(const Duration(milliseconds: 1500));
    if (mounted) setState(() => _showPulse = false);
  }

  Future<void> _updateTimed(double delta) async {
    final provider = context.read<HabitProvider>();
    final newVal = (_timedMinutes + delta).clamp(0, 999).toDouble();
    setState(() => _timedMinutes = newVal);
    await provider.updateTimedEntry(_habit, newVal, date: _targetDate);
    setState(() => _isCompleted = _habit.entryFor(_targetDate)?.isCompleted ?? newVal >= _habit.dailyTarget);
  }

  Future<void> _updateCount(double delta) async {
    final provider = context.read<HabitProvider>();
    final newVal = (_countValue + delta).clamp(0, 9999).toDouble();
    setState(() => _countValue = newVal);
    await provider.updateCountEntry(_habit, newVal, date: _targetDate);
    setState(() => _isCompleted = _habit.entryFor(_targetDate)?.isCompleted ?? newVal >= _habit.dailyTarget);
  }

  @override
  Widget build(BuildContext context) {
    final isPulse = context.select<HabitProvider, bool>(
      (p) => p.checkInPulseHabitId == _habit.id,
    );
    final isAbstain = _habit.habitTrackingType == HabitTrackingType.abstain;
    final accentColor = isAbstain ? TributeColor.sage : TributeColor.golden;

    return Stack(
      children: [
        GestureDetector(
          onTap: () => _showDetail(context),
          child: Container(
            padding: const EdgeInsets.all(16),
            decoration: TributeDecorations.card,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                _header(accentColor),
                const SizedBox(height: 12),
                _trackingUI(accentColor),
                if (_isCompleted && _completionVerse != null) ...[
                  const SizedBox(height: 12),
                  _verseSection(),
                ],
                if (isAbstain && !widget.isRetroactive) ...[
                  const SizedBox(height: 10),
                  _sosLink(),
                ],
              ],
            ),
          ),
        ),
        if (_showPulse || isPulse)
          Positioned.fill(
            child: IgnorePointer(
              child: Center(
                child: GoldenPulseView(onComplete: () {
                  if (mounted) setState(() => _showPulse = false);
                }),
              ),
            ),
          ),
        if (_celebrationMilestone != null)
          Positioned.fill(
            child: MilestoneCelebrationView(
              milestone: _celebrationMilestone!,
              onDismiss: () => setState(() => _celebrationMilestone = null),
            ),
          ),
      ],
    );
  }

  Widget _header(Color accentColor) {
    return Row(
      children: [
        Container(
          width: 40,
          height: 40,
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            gradient: RadialGradient(
              colors: [
                accentColor.withValues(alpha: _isCompleted ? 0.3 : 0.12),
                accentColor.withValues(alpha: _isCompleted ? 0.1 : 0.03),
              ],
            ),
          ),
          child: Icon(
            _habitIcon(),
            color: accentColor,
            size: 18,
          ),
        ),
        const SizedBox(width: 12),
        Expanded(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Text(
                _habit.name,
                style: const TextStyle(
                  fontWeight: FontWeight.w600,
                  fontSize: 15,
                  color: TributeColor.warmWhite,
                ),
              ),
              if (_isCompleted)
                Text(
                  _completedSubtitle(),
                  style: TextStyle(
                    fontSize: 11,
                    color: TributeColor.sage,
                  ),
                )
              else
                Text(
                  _habit.purposeStatement,
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                  style: TextStyle(
                    fontSize: 11,
                    color: TributeColor.softGold.withValues(alpha: 0.6),
                  ),
                ),
            ],
          ),
        ),
        if (_isCompleted)
          Icon(Icons.check_circle_rounded, color: accentColor, size: 24)
        else
          Icon(Icons.chevron_right, color: Colors.white.withValues(alpha: 0.2), size: 16),
      ],
    );
  }

  Widget _trackingUI(Color accentColor) {
    switch (_habit.habitTrackingType) {
      case HabitTrackingType.checkIn:
        return _checkInButton(accentColor);
      case HabitTrackingType.abstain:
        return _abstainButton();
      case HabitTrackingType.timed:
        return _timedUI(accentColor);
      case HabitTrackingType.count:
        return _countUI(accentColor);
    }
  }

  Widget _checkInButton(Color accentColor) {
    if (_isCompleted) return const SizedBox.shrink();
    return SizedBox(
      width: double.infinity,
      child: ElevatedButton(
        onPressed: _checkIn,
        style: ElevatedButton.styleFrom(
          backgroundColor: accentColor,
          foregroundColor: TributeColor.charcoal,
          padding: const EdgeInsets.symmetric(vertical: 12),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12)),
        ),
        child: const Text('Check In', style: TextStyle(fontWeight: FontWeight.w600)),
      ),
    );
  }

  Widget _sosLink() {
    return GestureDetector(
      onTap: () => _launchSOS(context),
      behavior: HitTestBehavior.opaque,
      child: Padding(
        padding: const EdgeInsets.symmetric(vertical: 2),
        child: Row(mainAxisAlignment: MainAxisAlignment.center, children: [
          Icon(Icons.shield_rounded, size: 13, color: TributeColor.warmCoral.withValues(alpha: 0.65)),
          const SizedBox(width: 5),
          Text('SOS — Need help right now?',
              style: TextStyle(fontSize: 12, color: TributeColor.warmCoral.withValues(alpha: 0.65))),
        ]),
      ),
    );
  }

  void _launchSOS(BuildContext context) {
    final isPremium = context.read<StoreProvider>().isPremium;
    if (!isPremium) {
      showModalBottomSheet(
        context: context,
        isScrollControlled: true,
        backgroundColor: TributeColor.charcoal,
        builder: (_) => const TributePaywallView(
          contextTitle: 'SOS Support',
          contextMessage: 'Tough moment? The SOS feature can help — it\'ll remind you why you started and connect you with your circle.',
        ),
      );
      return;
    }
    Navigator.of(context).push(
      MaterialPageRoute(
        fullscreenDialog: true,
        builder: (_) => SOSView(habit: _habit),
      ),
    );
  }

  String _dayName(DateTime date) {
    const days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'];
    return days[date.weekday - 1];
  }

  Widget _abstainButton() {
    if (_isCompleted) return const SizedBox.shrink();
    return SizedBox(
      width: double.infinity,
      child: ElevatedButton.icon(
        onPressed: _checkIn,
        icon: const Icon(Icons.shield_rounded, size: 16),
        label: Text(widget.isRetroactive
            ? 'Were you strong on ${_dayName(widget.targetDate)}?'
            : 'Stayed strong today?'),
        style: ElevatedButton.styleFrom(
          backgroundColor: TributeColor.sage,
          foregroundColor: TributeColor.charcoal,
          padding: const EdgeInsets.symmetric(vertical: 12),
          shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(12)),
        ),
      ),
    );
  }

  Widget _timedUI(Color accentColor) {
    final target = _habit.dailyTarget;
    final ratio = target > 0 ? (_timedMinutes / target).clamp(0.0, 1.0) : 0.0;

    return Column(
      children: [
        Stack(
          alignment: Alignment.center,
          children: [
            SizedBox(
              width: 80,
              height: 80,
              child: CircularProgressIndicator(
                value: ratio,
                strokeWidth: 6,
                backgroundColor: Colors.white.withValues(alpha: 0.08),
                valueColor: AlwaysStoppedAnimation(accentColor),
              ),
            ),
            Text(
              '${_timedMinutes.toInt()}m',
              style: TextStyle(
                fontSize: 16,
                fontWeight: FontWeight.w700,
                color: accentColor,
              ),
            ),
          ],
        ),
        const SizedBox(height: 12),
        Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            _timedButton('-5', () => _updateTimed(-5)),
            const SizedBox(width: 8),
            _timedButton('+5', () => _updateTimed(5)),
            const SizedBox(width: 8),
            _timedButton('+15', () => _updateTimed(15)),
            const SizedBox(width: 8),
            _timedButton('+30', () => _updateTimed(30)),
          ],
        ),
        if (target > 0) ...[
          const SizedBox(height: 6),
          Text(
            'Goal: ${target.toInt()} min',
            style: TextStyle(fontSize: 10, color: Colors.white.withValues(alpha: 0.3)),
          ),
        ],
      ],
    );
  }

  Widget _timedButton(String label, VoidCallback onTap) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 8),
        decoration: BoxDecoration(
          color: TributeColor.surfaceOverlay,
          borderRadius: BorderRadius.circular(8),
        ),
        child: Text(
          label,
          style: const TextStyle(fontSize: 12, fontWeight: FontWeight.w600, color: TributeColor.softGold),
        ),
      ),
    );
  }

  Widget _countUI(Color accentColor) {
    final target = _habit.dailyTarget;
    final unit = _habit.targetUnit.isEmpty ? '' : ' ${_habit.targetUnit}';
    return Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        _countButton(Icons.remove, () => _updateCount(-1)),
        const SizedBox(width: 16),
        Column(
          children: [
            Text(
              '${_countValue.toInt()}$unit',
              style: TextStyle(
                fontSize: 24,
                fontWeight: FontWeight.w700,
                color: _isCompleted ? accentColor : TributeColor.warmWhite,
              ),
            ),
            if (target > 0)
              Text(
                'of ${target.toInt()}$unit',
                style: TextStyle(fontSize: 10, color: Colors.white.withValues(alpha: 0.3)),
              ),
          ],
        ),
        const SizedBox(width: 16),
        _countButton(Icons.add, () => _updateCount(1)),
      ],
    );
  }

  Widget _countButton(IconData icon, VoidCallback onTap) {
    return GestureDetector(
      onTap: onTap,
      child: Container(
        width: 36,
        height: 36,
        decoration: BoxDecoration(
          shape: BoxShape.circle,
          color: TributeColor.surfaceOverlay,
        ),
        child: Icon(icon, size: 18, color: TributeColor.softGold),
      ),
    );
  }

  Widget _verseSection() {
    final verse = _completionVerse!;
    return Column(
      children: [
        Text(
          '\u201C${verse.text}\u201D',
          textAlign: TextAlign.center,
          style: TextStyle(
            fontSize: 11,
            fontStyle: FontStyle.italic,
            color: TributeColor.softGold.withValues(alpha: 0.55),
            height: 1.5,
          ),
        ),
        const SizedBox(height: 2),
        Text(
          verse.reference,
          style: TextStyle(fontSize: 10, color: TributeColor.golden.withValues(alpha: 0.4)),
        ),
      ],
    );
  }

  String _completedSubtitle() {
    switch (_habit.habitTrackingType) {
      case HabitTrackingType.timed:
        return '${_timedMinutes.toInt()} min given';
      case HabitTrackingType.count:
        final unit = _habit.targetUnit.isEmpty ? '' : ' ${_habit.targetUnit}';
        return '${_countValue.toInt()}$unit completed';
      case HabitTrackingType.checkIn:
        return '${_habit.totalCompletedDays()} days total';
      case HabitTrackingType.abstain:
        return 'Clean day \u2713';
    }
  }

  IconData _habitIcon() {
    if (_habit.habitTrackingType == HabitTrackingType.abstain) return Icons.shield_rounded;
    switch (_habit.habitCategory) {
      case HabitCategory.gratitude: return Icons.auto_awesome;
      case HabitCategory.scripture: return Icons.menu_book;
      case HabitCategory.exercise: return Icons.fitness_center;
      case HabitCategory.rest: return Icons.bedtime;
      case HabitCategory.fasting: return Icons.no_food;
      case HabitCategory.study: return Icons.school;
      case HabitCategory.service: return Icons.volunteer_activism;
      case HabitCategory.connection: return Icons.people;
      case HabitCategory.health: return Icons.favorite;
      case HabitCategory.abstain: return Icons.shield_rounded;
      case HabitCategory.custom: return Icons.star;
    }
  }

  void _showDetail(BuildContext context) {
    showModalBottomSheet(
      context: context,
      isScrollControlled: true,
      backgroundColor: TributeColor.charcoal,
      shape: const RoundedRectangleBorder(
        borderRadius: BorderRadius.vertical(top: Radius.circular(20)),
      ),
      builder: (_) => DraggableScrollableSheet(
        initialChildSize: 0.9,
        maxChildSize: 0.95,
        minChildSize: 0.5,
        expand: false,
        builder: (ctx, sc) => HabitDetailView(habit: _habit, scrollController: sc),
      ),
    );
  }
}
